src/main/scala/is/hail/methods/LinearMixedRegression.scala

```diff
@@ -184,97 +181,108 @@ object LinearMixedRegression {
+
+    Annotation(b, s2, chi2, p)
+  }
+}
+
+// Handles low-rank case, but is slower than ScalerLMM on full-rank case
+class LowRankScalerLMM(lmmConstants: LMMConstants, delta: Double, logNullS2: Double, useML: Boolean) extends ScalerLMM {
   val n = lmmConstants.n
   val d = lmmConstants.d
   val k = lmmConstants.S.length
   val y = lmmConstants.y
   val covs = lmmConstants.C
   val D = lmmConstants.S + delta
-  val Ut = lmmConstants.U.t
+  val Ut = lmmConstants.Ut
   val Uty = lmmConstants.Uty
   val UtCov = lmmConstants.UtC
   val yty = lmmConstants.yty
   val dy = Uty :/ D
-  val covTcov = lmmConstants.CtC
-
-  val covTy = lmmConstants.Cty
-
-  val Z = D.map(x => 1 / x) - 1 / delta
-  val ydy = yty / delta +
-    (Uty dot (Uty :* Z))
+  val covtcov = lmmConstants.CtC
+  val covty = lmmConstants.Cty
+  val invDelta = 1 / delta
+  val Z = D.map(1 / _ - invDelta)
+  val ydy = yty / delta + (Uty dot (Uty :* Z))
 
   val ZUtCov = UtCov(::, *) :* Z
   val CzCLowerRight = UtCov.t * ZUtCov
 
-  def likelihoodRatioTest(v: Vector[Double]): Annotation = {
-    val vty = v dot y
-    val vtv = v dot v
-    val covtv = covs.t * v
-    val Utv = Ut * v
-
-    val UtC = DenseMatrix.horzcat(Utv.toDenseVector.toDenseMatrix.t, UtCov)
-
-    var CtC = DenseMatrix.zeros[Double](d + 1, d + 1)
+  def likelihoodRatioTest(x: Vector[Double]): Annotation = {
+    val xty = x dot y
+    val xtx = x dot x
+    val covtx = covs.t * x
+    val Utx = Ut * x
+    val UtC = DenseMatrix.horzcat(Utx.toDenseVector.toDenseMatrix.t, UtCov)
+
+    val CtC = DenseMatrix.zeros[Double](d + 1, d + 1)
+    CtC(0, 0) = xtx
+    CtC(0 to 0, 1 to d) := covtx
+    CtC(1 to d, 0) := covtx
     CtC(1 to d, 1 to d) := lmmConstants.CtC
-    CtC(0, 0) = vtv
-
-    var i = 1
-    while (i <= d) {
-      CtC(0, i) = covtv(i - 1)
-      CtC(i, 0) = covtv(i - 1)
-      i += 1
-    }
-
-    val Cty = DenseVector.vertcat(DenseVector(vty), covTy)
 
-    val Cdy = Cty / delta +
-      (UtC.t  * (Uty :* Z))
-
-    val ZUtv = Utv :* Z
+    val Cty = DenseVector.vertcat(DenseVector(xty), covty)
+    val Cdy = invDelta * Cty + (UtC.t  * (Uty :* Z))
+    val ZUtx = Utx :* Z
 
     //4 Parts (lowerLeft and upperRight same)
-    val CzCUpperLeft = Utv dot ZUtv
-    val CzCLowerLeft = UtCov.t * ZUtv
-
-    var CzC = DenseMatrix.zeros[Double](d + 1, d + 1)
-    CzC(1 to d, 1 to d) := CzCLowerRight
+    val CzCUpperLeft = Utx dot ZUtx
+    val CzCLowerLeft = UtCov.t * ZUtx
 
+    val CzC = DenseMatrix.zeros[Double](d + 1, d + 1)
     CzC(0, 0) = CzCUpperLeft
-    i = 1
-    while (i <= d) {
-      CzC(i, 0) = CzCLowerLeft(i - 1)
-      CzC(0, i) = CzCLowerLeft(i - 1)
-      i += 1
-    }
-
-    val CdC = CtC / delta + CzC
+    CzC(0 to 0, 1 to d) := CzCLowerLeft
+    CzC(1 to d, 0) := CzCLowerLeft
+    CzC(1 to d, 1 to d) := CzCLowerRight
 
+    val CdC = invDelta * CtC + CzC
     val b = CdC \ Cdy
-
     val r2 = ydy - (Cdy dot b)
-
     val s2 = r2 / (if (useML) n else n - d)
```

Comment:
move `val invDof = 1 / (if (useML) n else n - Qt.rows)` outside the test. this stands for inverse degrees of freedom.  Kill r2 line, just `val s2 = invDof * (ydy - (Cdy dot b))`